framework/src/media/audio/audio_manager.c

```diff
@@ -504,7 +521,38 @@ int start_audio_stream_in(void *data, uint32_t frames)
 	}
 
 	do {
-		ret = pcm_readi(pcm, data, frames);
+		if (cur_card->resample.necessary) {
+			unsigned int frames_to_read = (unsigned int)((float)frames * cur_card->resample.ratio);
+			if (frames_to_read > (get_input_frame_count() * cur_card->resample.ratio)) {
+				frames_to_read = (get_input_frame_count() * cur_card->resample.ratio);
+			}
+
+			audvdbg("Read frame for resample: %u to %u\n", frames_to_read, frames);
+
+			ret = pcm_readi(pcm, cur_card->resample.readbuffer, frames_to_read);
+
+			src_data_t srcData;
```

Comment:
because of calling pcm_prepare to resolve underrun, below shoud be moved to outside of while loop
```
do {
    if (resample) {
         readi(pcm, data, modified-frames);
    } else {
         readi(pcm_data, frames);
    }
} while ();
resample_logic_here;
```

Reply:
I agree. But this modification work is a little heavy. I can modify the logic in another PR.